sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.json
+
+import org.apache.spark.sql.catalyst.{InternalRow, StructFilters}
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.types.StructType
+
+/**
+ * The class provides API for applying pushed down source filters to rows with
+ * struct schema parsed from JSON records. It assumes that:
+ *   1. reset() is called before any skipRow() calls for new row.
+ *   2. skipRow() can be called for any valid index of the struct fields,
+ *      and in any order.
+ *
+ * @param filters The pushed down source filters. The filters should refer to
+ *                the fields of the provided schema.
+ * @param schema The required schema of records from datasource files.
+ */
+class JsonFilters(filters: Seq[sources.Filter], schema: StructType)
+  extends StructFilters(filters, schema) {
+
+  /**
+   * Stateful JSON predicate that keeps track of its dependent references in the
+   * current row via `refCount`.
+   *
+   * @param predicate The predicate compiled from pushed down source filters.
+   * @param totalRefs The total amount of all filters references which the predicate
+   *                  compiled from.
+   * @param refCount The current number of predicate references in the row that have
+   *                 been not set yet. When `refCount` reaches zero, the predicate
+   *                 has all dependencies are set, and can be applied to the row.
+   */
+  case class JsonPredicate(predicate: BasePredicate, totalRefs: Int, var refCount: Int) {
```

Comment:
We might have to avoid using `var` in a case class (https://github.com/databricks/scala-style-guide#case-classes-and-immutability)

Reply:
It is not problem here. I can replace it by a regular class. This will just add boilerplate code - need to implement a copy for the class.

Reply:
I converted case class to a regular class locally, here is the patch:
```diff
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonFilters.scala
@@ -48,10 +48,13 @@ class JsonFilters(filters: Seq[sources.Filter], schema: StructType)
    *                 been not set yet. When `refCount` reaches zero, the predicate
    *                 has all dependencies are set, and can be applied to the row.
    */
-  case class JsonPredicate(predicate: BasePredicate, totalRefs: Int, var refCount: Int) {
+  class JsonPredicate(val predicate: BasePredicate, val totalRefs: Int, var refCount: Int) {
     def reset(): Unit = {
       refCount = totalRefs
     }
+    def withTotalRefs(amount: Int): JsonPredicate = {
+      new JsonPredicate(predicate, amount, refCount)
+    }
   }

   // Predicates compiled from the pushed down filters, and indexed by their references
@@ -75,14 +78,14 @@ class JsonFilters(filters: Seq[sources.Filter], schema: StructType)
         // Combine all filters from the same group by `And` because all filters should
         // return `true` to do not skip a row. The result is compiled to a predicate.
         .map { case (refSet, refsFilters) =>
-          (refSet, JsonPredicate(toPredicate(refsFilters), refSet.size, 0))
+          (refSet, new JsonPredicate(toPredicate(refsFilters), refSet.size, 0))
         }
       // Apply predicates w/o references like AlwaysTrue and AlwaysFalse to all fields.
       // We cannot set such predicates to a particular position because skipRow() can
       // be invoked for any index due to unpredictable order of JSON fields in JSON records.
       val withLiterals = groupedByRefSet.map { case (refSet, pred) =>
         if (refSet.isEmpty) {
-          (schema.fields.map(_.name).toSet, pred.copy(totalRefs = 1))
+          (schema.fields.map(_.name).toSet, pred.withTotalRefs(1))
         } else {
           (refSet, pred)
         }
```
Let me know if you are ok with that, I will push it in that case.

Reply:
how about
```
case class JsonPredicate(predicate: BasePredicate, totalRefs: Int) {
  var refCount: Int
}
```